File: .claude/skills/single-comic-prompt/scripts/generate-cover.py
#!/usr/bin/env python3
"""
Blog cover image generator — calls Gemini 3 Pro Image API.

Usage:
  python generate-cover.py --prompt "Your image prompt here"
  python generate-cover.py --prompt-file prompt.txt
  python generate-cover.py --prompt-file prompt.txt --reference reference.png
  python generate-cover.py --prompt-file prompt.txt --output my-cover.png --size 2K --aspect 16:9
"""

import argparse
import base64
import json
import os
import sys
import urllib.request
import urllib.error
from datetime import datetime
from pathlib import Path

API_BASE = "https://generativelanguage.googleapis.com/v1beta/models"
DEFAULT_MODEL = "gemini-3-pro-image-preview"
DEFAULT_ASPECT = "16:9"
DEFAULT_SIZE = "2K"
SKILL_DIR = Path(__file__).resolve().parent.parent
DEFAULT_OUTPUT_DIR = SKILL_DIR / "output"


def load_api_key():
    key = os.environ.get("GEMINI_API_KEY", "")
    if key:
        return key

    env_file = SKILL_DIR / ".env"
    if env_file.exists():
        for line in env_file.read_text().splitlines():
            line = line.strip()
            if line.startswith("GEMINI_API_KEY="):
                return line.split("=", 1)[1].strip().strip("\"'")

    print("ERROR: GEMINI_API_KEY not found. Set it as env var or in .env", file=sys.stderr)
    sys.exit(1)


def resolve_output_path(output_arg):
    if not output_arg:
        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        return DEFAULT_OUTPUT_DIR / f"{timestamp}-cover.png"

    output_path = Path(output_arg)
    if not output_path.is_absolute() and output_path.parent == Path("."):
        return DEFAULT_OUTPUT_DIR / output_path
    return output_path


def resolve_prompt_archive_path(output_path):
    return output_path.with_name(f"{output_path.stem}-prompt.txt")


def load_image_as_base64(path):
    data = Path(path).read_bytes()
    ext = Path(path).suffix.lower()
    mime = {".png": "image/png", ".jpg": "image/jpeg", ".jpeg": "image/jpeg", ".webp": "image/webp"}.get(ext, "image/png")
    return base64.b64encode(data).decode(), mime


def generate_image(
    prompt,
    reference_path=None,
    model=DEFAULT_MODEL,
    aspect_ratio=DEFAULT_ASPECT,
    image_size=DEFAULT_SIZE,
):
    api_key = load_api_key()
    url = f"{API_BASE}/{model}:generateContent"

    parts = []

    if reference_path:
        img_b64, mime = load_image_as_base64(reference_path)
        parts.append({"inlineData": {"mimeType": mime, "data": img_b64}})

    parts.append({"text": prompt})

    payload = {
        "contents": [{"parts": parts}],
        "generationConfig": {
            "responseModalities": ["IMAGE"],
            "imageConfig": {
                "aspectRatio": aspect_ratio,
                "imageSize": image_size,
            },
        },
    }

    req = urllib.request.Request(
        f"{url}?key={api_key}",
        data=json.dumps(payload).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            body = json.loads(resp.read())
    except urllib.error.HTTPError as e:
        error_body = e.read().decode()
        print(f"API error {e.code}: {error_body}", file=sys.stderr)
        sys.exit(1)

    for candidate in body.get("candidates", []):
        for part in candidate.get("content", {}).get("parts", []):
            inline = part.get("inlineData")
            if inline and inline.get("data"):
                return base64.b64decode(inline["data"])

    print("ERROR: No image data in API response.", file=sys.stderr)
    print(json.dumps(body, indent=2, ensure_ascii=False), file=sys.stderr)
    sys.exit(1)


def main():
    parser = argparse.ArgumentParser(description="Generate blog cover image via Gemini 3 Pro")
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--prompt", help="Image generation prompt text")
    group.add_argument("--prompt-file", help="Path to a text file containing the prompt")
    parser.add_argument("--reference", help="Path to a reference style image (optional)")
    parser.add_argument("--output", help="Output image path (default: output/<YYYYMMDD-HHMMSS>-cover.png; bare filenames are saved under output/)")
    parser.add_argument("--model", default=DEFAULT_MODEL, help=f"Model name (default: {DEFAULT_MODEL})")
    parser.add_argument("--aspect", default=DEFAULT_ASPECT, help=f"Aspect ratio (default: {DEFAULT_ASPECT})")
    parser.add_argument("--size", default=DEFAULT_SIZE, choices=["512", "1K", "2K", "4K"], help=f"Image size (default: {DEFAULT_SIZE})")

    args = parser.parse_args()

    if args.prompt_file:
        prompt = Path(args.prompt_file).read_text().strip()
    else:
        prompt = args.prompt

    if not prompt:
        print("ERROR: Prompt is empty.", file=sys.stderr)
        sys.exit(1)

    output_path = resolve_output_path(args.output)
    prompt_archive_path = resolve_prompt_archive_path(output_path)

    print(f"Model:   {args.model}")
    print(f"Size:    {args.size}")
    print(f"Aspect:  {args.aspect}")
    if args.reference:
        print(f"Ref img: {args.reference}")
    print(f"Output:  {output_path}")
    print(f"Prompt archive: {prompt_archive_path}")
    print(f"Prompt:  {prompt[:120]}{'...' if len(prompt) > 120 else ''}")
    print()
    print("Generating image...")

    image_bytes = generate_image(
        prompt=prompt,
        reference_path=args.reference,
        model=args.model,
        aspect_ratio=args.aspect,
        image_size=args.size,
    )

    output_path.parent.mkdir(parents=True, exist_ok=True)
    prompt_archive_path.write_text(prompt + "\n", encoding="utf-8")
    output_path.write_bytes(image_bytes)
    print(f"Done! Saved to {output_path} ({len(image_bytes)} bytes)")
    print(f"Prompt saved to {prompt_archive_path}")


if __name__ == "__main__":
    main()

File: .cursor/skills/comic-diagram-generation/scripts/generate-cover.mjs
#!/usr/bin/env node
/**
 * Comic explainer image generator - calls Gemini image API.
 *
 * Usage:
 *   node generate-cover.mjs --prompt "Your image prompt here"
 *   node generate-cover.mjs --prompt-file prompt.txt
 *   node generate-cover.mjs --prompt-file prompt.txt --reference reference.png
 *   node generate-cover.mjs --prompt-file prompt.txt --output my-cover.png --size 2K --aspect 16:9
 */

import { existsSync } from "node:fs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { spawnSync } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

const API_BASE = "https://generativelanguage.googleapis.com/v1beta/models";
const DEFAULT_MODEL = "gemini-3-pro-image-preview";
const DEFAULT_ASPECT = "16:9";
const DEFAULT_SIZE = "2K";
const VALID_SIZES = new Set(["512", "1K", "2K", "4K"]);

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const skillDir = path.resolve(__dirname, "..");

function printHelp() {
  console.log(`usage: generate-cover.mjs (--prompt PROMPT | --prompt-file PROMPT_FILE)
                          [--reference REFERENCE] [--output OUTPUT]
                          [--model MODEL] [--aspect ASPECT]
                          [--size 512|1K|2K|4K] [--proxy PROXY_URL]

Generate comic explainer image via Gemini image API

options:
  -h, --help            show this help message and exit
  --prompt PROMPT       Image generation prompt text
  --prompt-file FILE    Path to a text file containing the prompt
  --reference FILE      Path to a reference style image (optional)
  --output FILE         Output image path (default: docs/public/images/comic-<timestamp>.png)
  --model MODEL         Model name (default: ${DEFAULT_MODEL})
  --aspect ASPECT       Aspect ratio (default: ${DEFAULT_ASPECT})
  --size SIZE           Image size: 512, 1K, 2K, or 4K (default: ${DEFAULT_SIZE})
  --proxy PROXY_URL     Optional HTTP proxy, e.g. http://127.0.0.1:6984`);
}

function parseArgs(argv) {
  const args = {
    model: DEFAULT_MODEL,
    aspect: DEFAULT_ASPECT,
    size: DEFAULT_SIZE,
  };
  const valueOptions = new Set([
    "--prompt",
    "--prompt-file",
    "--reference",
    "--output",
    "--model",
    "--aspect",
    "--size",
    "--proxy",
  ]);

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];

    if (arg === "-h" || arg === "--help") {
      args.help = true;
      continue;
    }

    if (!valueOptions.has(arg)) {
      throw new Error(`Unknown option: ${arg}`);
    }

    const value = argv[i + 1];
    if (!value || value.startsWith("--")) {
      throw new Error(`Missing value for ${arg}`);
    }

    const key = arg.slice(2).replace(/-([a-z])/g, (_, char) => char.toUpperCase());
    args[key] = value;
    i += 1;
  }

  if (args.help) {
    return args;
  }

  if (Boolean(args.prompt) === Boolean(args.promptFile)) {
    throw new Error("Provide exactly one of --prompt or --prompt-file.");
  }

  if (!VALID_SIZES.has(args.size)) {
    throw new Error("--size must be one of: 512, 1K, 2K, 4K.");
  }

  return args;
}

async function loadApiKey() {
  if (process.env.GEMINI_API_KEY) {
    return process.env.GEMINI_API_KEY;
  }

  const envFile = path.join(skillDir, ".env");
  if (existsSync(envFile)) {
    const envText = await readFile(envFile, "utf8");
    for (const rawLine of envText.split(/\r?\n/)) {
      const line = rawLine.trim();
      if (line.startsWith("GEMINI_API_KEY=")) {
        return line.split("=", 2)[1].trim().replace(/^['"]|['"]$/g, "");
      }
    }
  }

  throw new Error("GEMINI_API_KEY not found. Set it as env var or in .env");
}

async function loadImageAsInlineData(filePath) {
  const data = await readFile(filePath);
  const ext = path.extname(filePath).toLowerCase();
  const mimeType =
    {
      ".png": "image/png",
      ".jpg": "image/jpeg",
      ".jpeg": "image/jpeg",
      ".webp": "image/webp",
    }[ext] ?? "image/png";

  return {
    inlineData: {
      mimeType,
      data: data.toString("base64"),
    },
  };
}

async function generateImage({ prompt, referencePath, model, aspectRatio, imageSize, proxy }) {
  const apiKey = await loadApiKey();
  const parts = [];

  if (referencePath) {
    parts.push(await loadImageAsInlineData(referencePath));
  }

  parts.push({ text: prompt });

  const payload = {
    contents: [{ parts }],
    generationConfig: {
      responseModalities: ["IMAGE"],
      imageConfig: {
        aspectRatio,
        imageSize,
      },
    },
  };

  const proxyUrl = proxy || process.env.HTTPS_PROXY || process.env.HTTP_PROXY || process.env.ALL_PROXY;
  if (proxyUrl) {
    return generateImageWithCurl({
      apiKey,
      model,
      payload,
      proxy: proxyUrl,
    });
  }

  const response = await fetch(`${API_BASE}/${model}:generateContent?key=${apiKey}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  const bodyText = await response.text();
  let body;
  try {
    body = JSON.parse(bodyText);
  } catch {
    throw new Error(`API returned non-JSON response (${response.status}): ${bodyText}`);
  }

  if (!response.ok) {
    throw new Error(`API error ${response.status}: ${JSON.stringify(body, null, 2)}`);
  }

  for (const candidate of body.candidates ?? []) {
    for (const part of candidate.content?.parts ?? []) {
      const inline = part.inlineData ?? part.inline_data;
      if (inline?.data) {
        return Buffer.from(inline.data, "base64");
      }
    }
  }

  throw new Error(`No image data in API response:\n${JSON.stringify(body, null, 2)}`);
}

function generateImageWithCurl({ apiKey, model, payload, proxy }) {
  const curl = process.platform === "win32" ? "curl.exe" : "curl";
  const result = spawnSync(
    curl,
    [
      "-sS",
      "--fail-with-body",
      "-x",
      proxy,
      "-X",
      "POST",
      "-H",
      "Content-Type: application/json",
      "--data-binary",
      "@-",
      `${API_BASE}/${model}:generateContent?key=${apiKey}`,
    ],
    {
      input: JSON.stringify(payload),
      encoding: "utf8",
      maxBuffer: 50 * 1024 * 1024,
    },
  );

  if (result.error) {
    throw result.error;
  }

  if (result.status !== 0) {
    throw new Error(`curl failed (${result.status}): ${result.stderr || result.stdout}`);
  }

  let body;
  try {
    body = JSON.parse(result.stdout);
  } catch {
    throw new Error(`API returned non-JSON response: ${result.stdout}`);
  }

  for (const candidate of body.candidates ?? []) {
    for (const part of candidate.content?.parts ?? []) {
      const inline = part.inlineData ?? part.inline_data;
      if (inline?.data) {
        return Buffer.from(inline.data, "base64");
      }
    }
  }

  throw new Error(`No image data in API response:\n${JSON.stringify(body, null, 2)}`);
}

function timestamp() {
  return new Date()
    .toISOString()
    .replace(/[-:]/g, "")
    .replace(/\..+$/, "")
    .replace("T", "-");
}

async function main() {
  const args = parseArgs(process.argv.slice(2));

  if (args.help) {
    printHelp();
    return;
  }

  const prompt = args.promptFile
    ? (await readFile(args.promptFile, "utf8")).trim()
    : args.prompt;

  if (!prompt) {
    throw new Error("Prompt is empty.");
  }

  const outputPath = args.output ?? path.join("docs", "public", "images", `comic-${timestamp()}.png`);

  console.log(`Model:   ${args.model}`);
  console.log(`Size:    ${args.size}`);
  console.log(`Aspect:  ${args.aspect}`);
  if (args.reference) {
    console.log(`Ref img: ${args.reference}`);
  }
  console.log(`Output:  ${outputPath}`);
  console.log(`Prompt:  ${prompt.slice(0, 120)}${prompt.length > 120 ? "..." : ""}`);
  console.log();
  console.log("Generating image...");

  const imageBytes = await generateImage({
    prompt,
    referencePath: args.reference,
    model: args.model,
    aspectRatio: args.aspect,
    imageSize: args.size,
    proxy: args.proxy,
  });

  await mkdir(path.dirname(outputPath), { recursive: true });
  await writeFile(outputPath, imageBytes);
  console.log(`Done! Saved to ${outputPath} (${imageBytes.length} bytes)`);
}

main().catch((error) => {
  console.error(`ERROR: ${error.message}`);
  process.exit(1);
});
